akka-projection-grpc/src/main/scala/akka/persistence/typed/EventWriter.scala

```diff
@@ -0,0 +1,328 @@
+                  case None =>
+                    if (context.log.isTraceEnabled)
+                      context.log.traceN(
+                        "Writing event persistence id [{}], sequence nr [{}], payload {}",
+                        persistenceId,
+                        sequenceNumber,
+                        event)
+                    sendToJournal(Vector(reprWithMeta))
+                    StateForPid(Map((reprWithMeta.sequenceNr, (reprWithMeta, replyTo))), emptyWaitingForWrite)
+                  case Some(state) =>
+                    // write in progress for pid, add write to batch and perform once current write completes
+                    if (state.waitingForWrite.size == maxBatchSize) {
+                      replyTo ! StatusReply.error(
+                        s"Max batch reached for pid $persistenceId, at most $maxBatchSize writes for " +
+                        "the same pid may be in flight at the same time")
+                      state
+                    } else {
+                      if (context.log.isTraceEnabled)
+                        context.log.traceN(
+                          "Writing event in progress for persistence id [{}], adding sequence nr [{}], payload {} to batch",
+                          persistenceId,
+                          sequenceNumber,
+                          event)
+                      state.copy(waitingForWrite = state.waitingForWrite :+ ((reprWithMeta, replyTo)))
+                    }
+                }
+              perPidWriteState = perPidWriteState.updated(persistenceId, newStateForPid)
+              Behaviors.same
+
+            case MaxSeqNrForPid(pid, maxSeqNr, originalErrorDesc) =>
+              // write failed, so we looked up the maxSeqNr to detect if it was duplicate events, already in journal
+              perPidWriteState.get(pid) match {
+                case None =>
+                  throw new IllegalStateException(
+                    s"MaxSeqNrForPid($pid, $maxSeqNr) returned but no such pid in state, this is a bug")
+                case Some(state) =>
+                  val sortedSeqs = state.waitingForReply.keys.toSeq.sorted
+                  val (alreadyInJournal, needsWrite) = sortedSeqs.partition(seqNr => seqNr <= maxSeqNr)
+                  if (alreadyInJournal.isEmpty) {
+                    // error was not about duplicates
+                    state.waitingForReply.values.foreach {
+                      case (_, replyTo) =>
+                        replyTo ! StatusReply.error("Journal write failed")
+                    }
+                    context.log.warnN(
+                      "Failed writing event batch persistence id [{}], sequence nr [{}-{}]: {}",
+                      pid,
+                      sortedSeqs.head,
+                      sortedSeqs.last,
+                      originalErrorDesc)
+                    val newState = state.copy(waitingForReply = Map.empty, writeErrorHandlingInProgress = false)
```

Comment:
Could it be possible that we haven't received all WriteMessageFailure (from previous batch) when we now change writeErrorHandlingInProgress, which means that we will trigger a new check for an old failure? Would be nice if we could drain all expected WriteMessageFailure before we trigger the check. Maybe a counter in the state, and we know how many to expect?

Reply:
Yes, good point, I think we have to deal with that. 

I started moving the writer over to Akka in https://github.com/akka/akka/pull/32010
Will take a look at this problem next over there.

Reply:
For the record I adressed all feedback here over in that PR.